### PR TITLE
Added DockerFile for docker containerization

### PR DESCRIPTION
**Pull request by johnmaxwilson**

### CODE REVIEW

**Comment by tjesser-ucdavis-edu** [comment]
3: That doesn't sound like a good fix, tbh. Mixing different geographic versions (1.45 vs 1.48) is bound to cause difficult-to-find issues, and if this ends up as the build guidelines for Virtual Quake, a dependency like this will make it unfeasible to build anywhere else.

2: I was able to find various resources talking about this problem, but haven't had time to sort through them. Here are a ~~few~~ lot of links:
- https://docs.docker.com/engine/reference/commandline/volume_create/#driver-specific-options
- https://stackoverflow.com/questions/32397496/docker-replicate-uid-gid-in-container-from-host
- https://stackoverflow.com/questions/43943251/docker-get-host-username-from-uid-via-shared-pid-namespace
- https://stackoverflow.com/questions/27925006/using-host-environment-variables-with-dockerfile

And some about shared volumes:
- https://stackoverflow.com/questions/23544282/what-is-the-best-way-to-manage-permissions-for-docker-shared-volumes
- https://stackoverflow.com/questions/22049212/docker-copy-file-from-container-to-host
- https://stackoverflow.com/questions/22907231/copying-files-from-host-to-docker-container
- https://github.com/moby/moby/issues/17798

**Comment by johnmaxwilson** [comment]
New changes!  

3: I've finally gotten cmake to identify the apt-get installed geographiclib, so that issue is now fixed.  We no longer need the geographiclib source.

2: I'm now setting the uid and gid of the virtualquake user via an entrypoint script.  For Linux users, the host uid and gid are passed as environment variables to the container through `docker run`; Docker for Mac seems to handle permissions between the container and the host file system differently, and they don't need to pass in the variables.
For Linux: `docker run --rm -v <full path to host directory>:/home/virtualquake/external_volume -e HOST_UID=$(id -u) -e HOST_GID=$(id -g) -it vq`
For Mac: `docker run --rm -v <full path to host directory>:/home/virtualquake/external_volume  -it vq`
The only issue that arose was that the entrypoint script is run as whatever user is last set in the Dockerfile, which means the container has to start up as root to have permission to change ids around.  My solution was to end the entrypoint script with a `su virtualquake`.  When the user actually gains control of the container terminal, they're doing so as virtualquake.  Does this seem appropriate?

File: docker/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+if [ -n "${HOST_UID}" ]
+  then
+    groupmod --gid=$HOST_GID virtualquake
+    usermod --uid=$HOST_UID virtualquake
+fi
+chown -R virtualquake:virtualquake /home/virtualquake/
+cd /home/virtualquake/external_volume
+su virtualquake
```
Comment (johnmaxwilson) [inline docker/entrypoint.sh]: I've played with a few options for running su and getting the sub shell to move to the right directory, but they have issues like starting multiple nested shells; I'm not really familiar with what problems carrying the root environment to the user could produce, but if it's not critical, the existing `su virtualquake` produces the behavior I want and so I'm good keeping it as is.

File: docker/Dockerfile [hunk 2]
```diff
@@ -23,61 +24,27 @@ RUN apt-get update && apt-get install -y \
...
+    -DGeographicLib_LIBRARY_DIRS="/usr/lib/x86_64-linux-gnu/" .. \
+  && cmake --build . -- all  \
+  && cmake --build . -- install \
+  && mkdir /home/virtualquake/external_volume
 
-WORKDIR /home/virtualquake/external_volume
+COPY entrypoint.sh /etc
 
-CMD [ "/bin/bash" ]
+ENTRYPOINT ["/bin/sh", "/etc/entrypoint.sh"]
```
Comment (johnmaxwilson) [inline docker/Dockerfile]: Fixed, changed Dockerfile to bash

File: docker/Dockerfile [hunk 3]
```diff
@@ -23,61 +24,27 @@ RUN apt-get update && apt-get install -y \
...
-RUN cd build \
-  && cmake --build . -- install
-
-USER virtualquake
-
-WORKDIR /home/virtualquake
-
-RUN mkdir external_volume
+  && cmake -DGeographicLib_INCLUDE_DIRS="/usr/include/GeographicLib" \
+    -DGeographicLib_LIBRARY_DIRS="/usr/lib/x86_64-linux-gnu/" .. \
```
Comment (johnmaxwilson) [inline docker/Dockerfile]: Agreed, put each flag on a separate line.

File: docker/entrypoint.sh (hunk 1 above)
Comment (tjesser-ucdavis-edu) [inline docker/entrypoint.sh]: Works for me.

**Comment by tjesser-ucdavis-edu** [comment]
You seem to have included an unrelated commit, https://github.com/geodynamics/vq/pull/191/commits/1bd88115501b217861bd2f5dd392bfa63cbb7aaa, but other than that, looks good. If you know how to use the necessary git incantations, it would probably be worth fixing, otherwise it's up to you. I'll work on getting this onto dockerhub soon.

**Comment by johnmaxwilson** [comment]
Thanks for catching that.  Removed that commit, and merging.

**Pull request merged**